pkg/server/plugin/keymanager/hashicorpvault/hashicorp_vault.go

```diff
@@ -0,0 +1,490 @@
+	}, nil
+}
+
+func (p *Plugin) GetPublicKeys(context.Context, *keymanagerv1.GetPublicKeysRequest) (*keymanagerv1.GetPublicKeysResponse, error) {
+	var keys = make([]*keymanagerv1.PublicKey, 0, len(p.entries))
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	for _, key := range p.entries {
+		keys = append(keys, key.PublicKey)
+	}
+
+	return &keymanagerv1.GetPublicKeysResponse{PublicKeys: keys}, nil
+}
+
+func algosForKMS(keyType keymanagerv1.KeyType, signerOpts any) (TransitHashAlgorithm, TransitSignatureAlgorithm, error) {
+	var (
+		hashAlgo keymanagerv1.HashAlgorithm
+		isPSS    bool
+	)
+
+	switch opts := signerOpts.(type) {
+	case *keymanagerv1.SignDataRequest_HashAlgorithm:
+		hashAlgo = opts.HashAlgorithm
+		isPSS = false
+	case *keymanagerv1.SignDataRequest_PssOptions:
+		if opts.PssOptions == nil {
+			return "", "", errors.New("PSS options are required")
+		}
+		hashAlgo = opts.PssOptions.HashAlgorithm
+		isPSS = true
+		// opts.PssOptions.SaltLength is handled by Vault. The salt length matches the bits of the hashing algorithm.
+	default:
+		return "", "", fmt.Errorf("unsupported signer opts type %T", opts)
+	}
+
+	isRSA := keyType == keymanagerv1.KeyType_RSA_2048 || keyType == keymanagerv1.KeyType_RSA_4096
+
+	switch {
+	case hashAlgo == keymanagerv1.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM:
+		return "", "", errors.New("hash algorithm is required")
+	case keyType == keymanagerv1.KeyType_EC_P256 || keyType == keymanagerv1.KeyType_EC_P384:
+		return TransitHashAlgorithmNone, TransitSignatureSignatureAlgorithmPKCS1v15, nil
+	case isRSA && !isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA256:
+		return TransitHashAlgorithmSHA256, TransitSignatureSignatureAlgorithmPKCS1v15, nil
+	case isRSA && !isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA384:
+		return TransitHashAlgorithmSHA384, TransitSignatureSignatureAlgorithmPKCS1v15, nil
+	case isRSA && !isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA512:
+		return TransitHashAlgorithmSHA512, TransitSignatureSignatureAlgorithmPKCS1v15, nil
+	case isRSA && isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA256:
+		return TransitHashAlgorithmSHA256, TransitSignatureSignatureAlgorithmPSS, nil
+	case isRSA && isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA384:
+		return TransitHashAlgorithmSHA384, TransitSignatureSignatureAlgorithmPSS, nil
+	case isRSA && isPSS && hashAlgo == keymanagerv1.HashAlgorithm_SHA512:
+		return TransitHashAlgorithmSHA512, TransitSignatureSignatureAlgorithmPSS, nil
+	default:
+		return "", "", fmt.Errorf("unsupported combination of keytype: %v and hashing algorithm: %v", keyType, hashAlgo)
+	}
+}
+
+func (p *Plugin) createKey(ctx context.Context, spireKeyID string, keyType keymanagerv1.KeyType) (*keyEntry, error) {
+	if p.vc == nil {
+		err := p.genVaultClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kt, err := convertToTransitKeyType(keyType)
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.vc.CreateKey(ctx, spireKeyID, *kt)
```

Comment:
What happens if the key already exists in the specified path?
I'm worried about only the possibility of creating a new key the only the first time, and then always using the same key.